### PR TITLE
[fix][broker] Reject auto create partitioned topic when topic name contains ``-partition-``

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

When the user set config as follow:

```json
{
    allowAutoTopicCreation: true,
    allowAutoTopicCreationType: "partitioned",
    defaultNumPartitions: 1
}
```
they can create partitioned topic success with topic name ``persistent://prop/autoNs/failedcreate-partition-abcde``. 

run get-partitioned-topic-metadata

```json
{
  "partitions" : 1
}
```
run ``admin.topics().deletePartitionedTopic(topicName);``

> Partitioned Topic Name should not contain '-partition-'

run ``admin.topics().delete(topicName)``

this operation can success delete the topic, but metadata still exists.

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [x] Make sure that the change passes the CI checks.

Add TCP/HTTP client test for this change.

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java [hunk 1]
```diff
@@ -31,26 +31,24 @@
...
 import org.apache.bookkeeper.client.LedgerHandle;
 import org.apache.bookkeeper.mledger.ManagedLedger;
 import org.apache.pulsar.broker.service.BrokerTestBase;
-import org.apache.pulsar.client.api.Consumer;
-import org.apache.pulsar.client.api.Message;
-import org.apache.pulsar.client.api.MessageRoutingMode;
-import org.apache.pulsar.client.api.Producer;
-import org.apache.pulsar.client.api.Schema;
-import org.apache.pulsar.client.api.SubscriptionType;
+import org.apache.pulsar.client.api.*;
```
Comment (Technoboy-) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:45]: Should not use `*`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 2]
```diff
@@ -2606,40 +2606,30 @@ private void createPendingLoadTopic() {
...
+                                    && !topicExists
+                                    && !topicName.isPartitioned()
+                                    && pulsar.getBrokerService().isAllowAutoTopicCreation(topicName)
+                                    && pulsar.getBrokerService().isDefaultTopicTypePartitioned(topicName)) {
+                                return pulsar.getBrokerService().createDefaultPartitionedTopicAsync(topicName);
+                            } else {
+                                return CompletableFuture.completedFuture(null);
+                            }
+                        }, pulsar.getExecutor())
+                );
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2624]: Why change line 2609~2624? Only for removing reducant `future` ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 3]
```diff
@@ -2606,40 +2606,30 @@ private void createPendingLoadTopic() {
...
+                            // MetadataStore callback thread.
+                            // If topic is already exist, creating partitioned topic is not allowed.
+                            if (metadata.partitions == 0
+                                    && !topicExists
+                                    && !topicName.isPartitioned()
+                                    && pulsar.getBrokerService().isAllowAutoTopicCreation(topicName)
+                                    && pulsar.getBrokerService().isDefaultTopicTypePartitioned(topicName)) {
+                                return pulsar.getBrokerService().createDefaultPartitionedTopicAsync(topicName);
+                            } else {
+                                return CompletableFuture.completedFuture(null);
```
Comment (AnonHxy) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2621]: Return `CompletableFuture.completedFuture(null)` will cause NPE by the methods who invoke it if  `metadata.partitions !=0`.  Maybe should return `CompletableFuture.completedFuture(metadata)` I think

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 2 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2624]: Yes, only for removing redundant future.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 1 above)
Comment (mattisonchao) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:45]: Sure, thanks~

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 3 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2621]: Good catch! thanks.

**Comment by mattisonchao** [comment]
@Technoboy-  @AnonHxy  PTAL :)

**Review by AnonHxy — APPROVED** [review]

**Comment by AnonHxy** [comment]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 4]
```diff
@@ -2606,40 +2606,30 @@ private void createPendingLoadTopic() {
...
+                        }, pulsar.getExecutor())
+                );
     }
 
     @SuppressWarnings("deprecation")
     private CompletableFuture<PartitionedTopicMetadata> createDefaultPartitionedTopicAsync(TopicName topicName) {
+        if (topicName.toString().contains(TopicName.PARTITIONED_TOPIC_SUFFIX)) {
+            return FutureUtil.failedFuture(new PulsarServerException.
+                    InvalidTopicNameException("Invalid topic name: " + topicName));
+        }
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2632]: The prompt message is not clear , like : `"Invalid topic name: " + topicName + ",  should not contain -partition-" `

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java [hunk 5]
```diff
@@ -271,4 +274,38 @@ public void testPersistentPartitionedTopicUnload() throws Exception {
...
+                    .topic(topicName)
+                    .create();
+            Assert.fail("unexpected operation");
+        } catch (PulsarClientException ex) {
+            Assert.assertTrue(ex.getMessage()
+                    .contains("Invalid topic name"));
+        }
+        Assert.assertEquals(admin.topics().getList(ns).size(), 0);
+        pulsar.getConfiguration().setAllowAutoTopicCreationType("non-partitioned");
+    }
```
Comment (RobertIndie) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:310]: Seems that we also need to add the `non-partitioned` type test case.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 4 above)
Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2632]: +1, We should give exact reason.

**Review by Jason918 — COMMENTED** [review]
I think it's better to narrow the limitation to ends with `-partition-\d+`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 6]
```diff
@@ -2606,40 +2606,30 @@ private void createPendingLoadTopic() {
...
-                                                });
-                                    } else {
-                                        future.complete(metadata);
-                                    }
-                                });
-
-                                return future;
-                            });
-                });
+                .thenCompose(topicExists -> fetchPartitionedTopicMetadataAsync(topicName)
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2609]: Seems not related to this PR?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 4 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2647]: fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 6 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java (hunk 5 above)
Comment (mattisonchao) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:310]: line 309 just to reset the configuration.

I'm not sure if clients should be denied creating a non-partitioned topic with the ``-partitioned-`` name, but I don't think it's a big deal, since a non-partitioned topic name containing ``-partitioned-`` doesn't seem to affect the logic.
I'm still wondering about that.
Thanks and let me know what you think.

Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentTopicTest.java:310]: Thanks for your explanation. Agree with you.

**Review by RobertIndie — APPROVED** [review]

**Review by Technoboy- — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 7]
```diff
@@ -2640,6 +2640,11 @@ private void createPendingLoadTopic() {
 
     @SuppressWarnings("deprecation")
     private CompletableFuture<PartitionedTopicMetadata> createDefaultPartitionedTopicAsync(TopicName topicName) {
+        if (topicName.toString().contains(TopicName.PARTITIONED_TOPIC_SUFFIX)) {
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:2643]: Can we create a namespace contains "-partition-" ?

**Comment by Jason918** [comment]
> I think it's better to narrow the limitation to ends with `-partition-\d+`

@mattisonchao What do you think about this?

**Comment by mattisonchao** [comment]
> > I think it's better to narrow the limitation to ends with `-partition-\d+`
> 
> @mattisonchao What do you think about this?

@Jason918 
I'm sorry for omitting this message, I think it's a great idea.

**Comment by mattisonchao** [comment]
Hi, @Jason918  
I use ``topicName.getLocalName()`` to solve the problem you mentioned. PTAL:)

**Review by Jason918 — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 8]
```diff
@@ -504,6 +505,9 @@ protected static PartitionedTopicMetadata fetchPartitionedTopicMetadataCheckAllo
             if (e.getCause() instanceof RestException) {
                 throw (RestException) e.getCause();
             }
+            if (e.getCause() instanceof PulsarServerException.InvalidTopicNameException) {
+                throw new RestException(Status.PRECONDITION_FAILED, e.getMessage());
+            }
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:510]: e.getMessage() or e.getCause().getMessage() ?

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:510]: fixed.

**Comment by codelipenghui** [comment]
@mattisonchao It might introduce a breaking change for DLQ, please help check the topic auto-creation with partitioned topic can work with DLQ with this change.

**Comment by mattisonchao** [comment]
Yes, We have to revert this PR and start a discussion in the mail list to find another graceful method.
